plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -1604,9 +1603,24 @@ private void flushBuffer(boolean finished)
             if (bufferSize == buffer.length || (finished && bufferSize > 0)) {
                 byte[] data = buffer;
                 int length = bufferSize;
-                this.buffer = new byte[buffer.length];
-                bufferSize = 0;
 
+                if (finished) {
+                    this.buffer = null;
```

Comment:
looks like unrelated optimization, move to separate commit

Reply:
Done. 
Also added a tiny `Make upload abort idempotent` commit.